File: python/mirage/workspace/session/permissions.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

from pydantic import BaseModel, ConfigDict, field_validator

from mirage.policy.types import DEFAULT_DENY_REASON, CommandRule
from mirage.types import HiddenPaths, HiddenVars, MountMode, parse_mount_mode

_DOC = ConfigDict(extra="forbid", frozen=True)

_RULE_FIELDS = frozenset({"reason", "commands", "paths"})


def _norm_prefix(prefix: str) -> str:
    """One spelling for a mount prefix: leading slash, no trailing one.

    Args:
        prefix (str): a prefix as typed in the document.
    """
    return "/" + prefix.strip("/")


def _list(value: Any, where: str, expected: str = "a list") -> tuple[Any, ...]:
    """A document list, refused before a scalar can be iterated.

    Args:
        value (Any): the field as written, None when absent.
        where (str): the field's name, for the message.
        expected (str): the expected shape named in the message.
    """
    if value is None:
        return ()
    if not isinstance(value, (list, tuple)):
        raise ValueError(f"{where} must be {expected}")
    return tuple(value)


def _string_list(value: Any, where: str) -> tuple[str, ...]:
    """A document list field, refused unless every item is a string.

    A scalar ``commands: rm`` would otherwise ``tuple()`` into
    ``('r', 'm')`` and the command it meant to refuse stay allowed, so
    the document fails to load instead, as it does in TypeScript.

    Args:
        value (Any): the field as written, None when absent.
        where (str): the field's name, for the message.
    """
    entries = _list(value, where, "a list of strings")
    for i, entry in enumerate(entries):
        if not isinstance(entry, str):
            raise ValueError(f"{where}[{i}] must be a string")
    return entries


def _rule(entry: Any) -> Any:
    """Coerce one ``deny`` entry to a CommandRule.

    A bare string is one command name with the default reason; a
    mapping is the rule's fields, ``reason`` defaulting; anything else
    is handed back for pydantic to report.

    Args:
        entry (Any): one entry as written in the document.
    """
    if isinstance(entry, str):
        return CommandRule(reason=DEFAULT_DENY_REASON, commands=(entry, ))
    if isinstance(entry, Mapping):
        unknown = sorted(set(entry) - _RULE_FIELDS)
        if unknown:
            raise ValueError(
                f"deny rule has unknown field(s): {', '.join(unknown)}")
        reason = entry.get("reason", DEFAULT_DENY_REASON)
        if not isinstance(reason, str):
            raise ValueError("deny rule reason must be a string")
        return CommandRule(reason=reason,
                           commands=_string_list(entry.get("commands"),
                                                 "deny rule commands"),
                           paths=_string_list(entry.get("paths"),
                                              "deny rule paths"))
    return entry


class PathsBlock(BaseModel):
    """``paths:`` of one tier.

    ``hide`` entries use the document's one grammar: an entry with
    ``*``, ``?`` or ``[`` is a pattern, anything else an exact path
    and its subtree (``utils/hidden.classify_paths``). ``show`` arrives
    with its enforcement.

    Args:
        hide (tuple[str, ...]): what the tier makes nonexistent.
    """

    model_config = _DOC

    hide: tuple[str, ...] = ()


class VarsBlock(BaseModel):
    """``vars:`` of a profile.

    Args:
        hide (tuple[str, ...]): variable names or globs over names the
            session reads as unset.
    """

    model_config = _DOC

    hide: tuple[str, ...] = ()


class CommandsBlock(BaseModel):
    """``commands:`` of the workspace tier.

    ``deny`` rules refuse with a reason; a bare string is one command
    name with the default reason. ``allow`` and ``ask`` arrive with
    their enforcement.

    Args:
        deny (tuple[CommandRule, ...]): the tier's refusals, in order.
    """

    model_config = _DOC

    deny: tuple[CommandRule, ...] = ()

    @field_validator("deny", mode="before")
    @classmethod
    def _v_deny(cls, v: Any) -> Any:
        return tuple(_rule(entry) for entry in _list(v, "commands.deny"))


class MountPermissions(BaseModel):
    """``mounts.<prefix>.permissions``: mount-owned, relative to the
    mount root, binding every session.

    Args:
        paths (PathsBlock): the mount's hides, mount-relative.
    """

    model_config = _DOC

    paths: PathsBlock = PathsBlock()


class WorkspacePermissions(BaseModel):
    """Top-level ``permissions:``: workspace-wide, absolute paths,
    binding every session.

    Args:
        commands (CommandsBlock): the workspace's deny rules.
        paths (PathsBlock): the workspace's hides.
    """

    model_config = _DOC

    commands: CommandsBlock = CommandsBlock()
    paths: PathsBlock = PathsBlock()


class SessionProfile(BaseModel):
    """One role's narrowing: the profile a session is created from, the
    inline document that tightens it, and the shape of both.

    Configuration, not enforcement: the resolver compiles the fields
    onto the session's own narrowing fields and the doors keep
    enforcing. A profile is a template (``extends`` is field
    inheritance: a stated field replaces the parent's, an absent one is
    inherited); safety comes from the layer intersection at
    evaluation, never from inheritance. Deliberately not named a View,
    which per the view convention is a door-scoped handle an agent
    holds, while a profile is what the embedder uses to *define* one.
    Frozen so two agents with the same role share one object and
    neither can bend the other's view. Every field is None when the
    document leaves it unsaid, which is what inheritance reads.

    Args:
        extends (str | None): the profile this one inherits from.
        cwd (str | None): the session's working directory at creation.
        env (dict[str, str] | None): a process environment seeded and
            exported into the session at creation.
        mounts (dict[str, MountMode] | tuple[str, ...] | None): a
            mapping of prefix to mode ceiling (``r`` / ``rw`` / ``rwx``
            or the mode names), or a plain list of prefixes that keeps
            each mount at its own configured mode; None leaves mounts
            unrestricted.
        paths (PathsBlock | None): the profile's hides.
        vars (VarsBlock | None): the profile's hidden variables.
    """

    model_config = _DOC

    extends: str | None = None
    cwd: str | None = None
    env: dict[str, str] | None = None
    mounts: dict[str, MountMode] | tuple[str, ...] | None = None
    paths: PathsBlock | None = None
    vars: VarsBlock | None = None

    @field_validator("mounts", mode="before")
    @classmethod
    def _v_mounts(cls, v: Any) -> Any:
        if v is None:
            return None
        if isinstance(v, str):
            return (_norm_prefix(v), )
        if isinstance(v, Mapping):
            modes: dict[str, MountMode] = {}
            for prefix, mode in v.items():
                if not isinstance(prefix, str):
                    raise ValueError("mounts keys must be strings")
                if not isinstance(mode, str):
                    raise ValueError(
                        f"mounts[{prefix}] must be a mode name or alias")
                modes[_norm_prefix(prefix)] = parse_mount_mode(mode)
            return modes
        if not isinstance(v, (list, tuple)):
            raise ValueError("mounts must be a mapping or a list of strings")
        return tuple(
            _norm_prefix(prefix) for prefix in _string_list(v, "mounts"))


@dataclass(frozen=True, slots=True)
class CompiledProfile:
    """The session fields an effective profile compiles to.

    Args:
        mount_modes (dict[str, MountMode] | None): per-mount ceilings;
            a listed prefix with no ceiling carries the mount's own
            mode as EXEC (no narrowing below the mount).
        hidden_paths (HiddenPaths | None): the profile's own hides.
        hidden_vars (HiddenVars | None): the profile's hidden variables.
        env (dict[str, str] | None): variables to seed and export.
        cwd (str | None): the working directory to start in.
    """

    mount_modes: dict[str, MountMode] | None
    hidden_paths: HiddenPaths | None
    hidden_vars: HiddenVars | None
    env: dict[str, str] | None
    cwd: str | None

File: typescript/packages/core/src/workspace/session/permissions.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { DEFAULT_ASK_REASON, DEFAULT_DENY_REASON } from '../../policy/constants.ts'
import type { CommandRule, CommandsSpec } from '../../policy/types.ts'
import type { HiddenPaths, HiddenVars } from '../../types.ts'
import { type MountMode, parseMountMode } from '../../types.ts'
import { isGlob } from '../../utils/hidden.ts'
import { stripSlash } from '../../utils/slash.ts'

/**
 * `paths:` of one tier. `hide` entries use the document's one grammar:
 * an entry with `*`, `?` or `[` is a pattern, anything else an exact
 * path and its subtree (`utils/hidden.classifyPaths`); every entry holds
 * a token, and the tier that owns the block decides whether it must be
 * absolute (workspace, profile) or is mount-relative (mount). `show`
 * arrives with its enforcement.
 */
export interface PathsBlock {
  readonly hide: readonly string[]
}

/** `vars:` of a profile: names or globs over names the session reads as unset. */
export interface VarsBlock {
  readonly hide: readonly string[]
}

/**
 * `commands:` of the workspace and profile tiers. `allow` lists the
 * command patterns the tier installs; a name none of them starts with is
 * not a command for the session (127, absent from `type` / `which` /
 * `man`), a line no pattern covers is refused. Grammar-tier shell
 * builtins and the agent's own functions are not subjects. `ask` rules
 * are admitted only with a host approval; `deny` rules refuse with a
 * reason. A bare string in either is one command pattern with the
 * default reason. `allow` null or absent (unstated) installs everything.
 */
export interface CommandsBlock {
  readonly allow?: readonly string[] | null
  readonly ask?: readonly CommandRule[]
  readonly deny: readonly CommandRule[]
}

/**
 * `commands:` of a mount tier: `ask` and `deny` only. A mount rule
 * applies to a line that works inside the mount (its cwd or one of its
 * paths lies under the root); its `paths` are mount-relative. There is
 * no mount-tier `allow`: what a session can see is a property of the
 * session, and an operand cannot make a command "not found".
 */
export interface MountCommandsBlock {
  readonly ask?: readonly CommandRule[]
  readonly deny: readonly CommandRule[]
}

/** `mounts.<prefix>.permissions`: mount-owned, relative to the mount root, binding every session. */
export interface MountPermissions {
  readonly paths: PathsBlock
  readonly commands?: MountCommandsBlock
}

/** Top-level `permissions:`: workspace-wide, absolute paths, binding every session. */
export interface WorkspacePermissions {
  readonly commands: CommandsBlock
  readonly paths: PathsBlock
}

/**
 * One role's narrowing: the profile a session is created from, the
 * inline document that tightens it, and the shape of both.
 *
 * Configuration, not enforcement: the resolver compiles the fields
 * onto the session's own narrowing fields and the doors keep
 * enforcing. A profile is a template (`extends` is field inheritance:
 * a stated field replaces the parent's, an absent one is inherited);
 * safety comes from the layer intersection at evaluation, never from
 * inheritance. Deliberately not named a View, which per the view
 * convention is a door-scoped handle an agent holds, while a profile
 * is what the embedder uses to *define* one. Immutable by type, so two
 * agents with the same role share one object and neither can bend the
 * other's view. Every field is absent/null when the document leaves it
 * unsaid, which is what inheritance reads.
 *
 * `mounts` is a Map or Record of prefix to mode ceiling (a mode name or
 * `r` / `rw` / `rwx`), or an array of prefixes that keeps each mount at
 * its own configured mode; `parseProfileMounts` normalizes every
 * spelling and the resolver reads only the normalized form.
 */
export interface SessionProfile {
  readonly extends?: string | null
  readonly cwd?: string | null
  readonly env?: Readonly<Record<string, string>> | null
  readonly mounts?:
    | ReadonlyMap<string, string>
    | Readonly<Record<string, string>>
    | readonly string[]
    | null
  readonly paths?: PathsBlock | null
  readonly vars?: VarsBlock | null
  readonly commands?: CommandsBlock | null
}

/**
 * The session fields an effective profile compiles to. `commands` is
 * the profile's own command tier, evaluated after the bound tiers.
 */
export interface CompiledProfile {
  readonly mountModes: ReadonlyMap<string, MountMode> | null
  readonly hiddenPaths: HiddenPaths | null
  readonly hiddenVars: HiddenVars | null
  readonly env: Readonly<Record<string, string>> | null
  readonly cwd: string | null
  readonly commands: CommandsSpec | null
}

const RULE_FIELDS = ['reason', 'commands', 'paths'] as const
const PATHS_FIELDS = ['hide'] as const
const VARS_FIELDS = ['hide'] as const
const COMMANDS_FIELDS = ['allow', 'ask', 'deny'] as const
const MOUNT_COMMANDS_FIELDS = ['ask', 'deny'] as const
const MOUNT_PERMISSIONS_FIELDS = ['paths', 'commands'] as const
const WORKSPACE_PERMISSIONS_FIELDS = ['commands', 'paths'] as const
const PROFILE_FIELDS = ['extends', 'cwd', 'env', 'mounts', 'paths', 'vars', 'commands'] as const

// A document mapping, not merely "an object": a Set, a Date or any class
// instance has no own enumerable string keys, so Object.entries would read
// one as an empty mapping and a `mounts` Set would compile to a session
// granting nothing at all. Python refuses the same values loudly.
function isPlainObject(v: unknown): v is Record<string, unknown> {
  if (typeof v !== 'object' || v === null) return false
  const proto: unknown = Object.getPrototypeOf(v)
  return proto === Object.prototype || proto === null
}

function rejectUnknownKeys(
  block: Record<string, unknown>,
  allowed: readonly string[],
  where: string,
): void {
  for (const key of Object.keys(block)) {
    if (!allowed.includes(key)) {
      throw new Error(`${where}: unknown field \`${key}\` (allowed: ${allowed.join(', ')})`)
    }
  }
}

function asObject(raw: unknown, where: string): Record<string, unknown> {
  if (!isPlainObject(raw)) throw new Error(`${where} must be a mapping`)
  return raw
}

/** A document list, refused before a scalar can be iterated (python's `_list`). */
function asList(raw: unknown, where: string, expected = 'a list'): readonly unknown[] {
  if (raw === undefined || raw === null) return []
  if (!Array.isArray(raw)) throw new Error(`${where} must be ${expected}`)
  return raw as readonly unknown[]
}

// An entry that names something must hold a token: a blank command
// pattern is a prefix of every line, so a stray "" would allow, ask about
// or deny every command, and a blank path entry is the root, so it would
// hide or deny the whole tree. `names` says what an entry names ("a
// command", "a path"); omitted, blank entries pass.
function stringList(raw: unknown, where: string, names?: string): readonly string[] {
  return asList(raw, where, 'a list of strings').map((entry, i) => {
    if (typeof entry !== 'string') throw new Error(`${where}[${String(i)}] must be a string`)
    if (names !== undefined && entry.trim() === '')
      throw new Error(`${where}[${String(i)}] must name ${names}`)
    return entry
  })
}

/**
 * Refuse a relative path entry where paths are absolute. The workspace
 * and profile tiers speak in virtual paths, so an entry there is either
 * absolute or a name pattern (`*.pem`, no slash, matching a path
 * component anywhere). A plain `xxx` or an anchored `secrets/*` would
 * otherwise be read from the root (`/xxx`, `/secrets/*`), which is never
 * what a relative spelling meant; the mount tier is the one place entries
 * are relative, to the mount root, and it is not checked here.
 */
function requireAbsolute(entries: readonly string[], where: string): void {
  entries.forEach((entry, i) => {
    if (entry.startsWith('/') || (isGlob(entry) && !entry.includes('/'))) return
    throw new Error(
      `${where}[${String(i)}] must be an absolute path or a name pattern at this tier: ` +
        `${JSON.stringify(entry)} is relative`,
    )
  })
}

function normPrefix(prefix: string): string {
  return '/' + stripSlash(prefix)
}

/**
 * The rules of a `commands` mapping: each command on its own paths, one
 * rule per entry, so the document never states a command beside a path
 * it was not meant for (`{rm: ['/repo/*'], mv: ['/shared/*']}` scopes
 * `rm` to the repo and `mv` to the share, nothing else).
 */
function scopedRules(
  commands: Record<string, unknown>,
  reason: string,
  where: string,
): CommandRule[] {
  const entries = Object.entries(commands)
  if (entries.length === 0) throw new Error(`${where}.commands must name at least one command`)
  return entries.map(([pattern, paths]) => {
    if (pattern.trim() === '') throw new Error(`${where}.commands keys must name a command`)
    const entries = stringList(paths, `${where}.commands[${pattern}]`, 'a path')
    if (entries.length === 0) {
      throw new Error(`${where}.commands[${pattern}] must list at least one path`)
    }
    return { reason, commands: [pattern], paths: entries }
  })
}

/**
 * Coerce one `deny` or `ask` entry to its rules. A bare string is one
 * command pattern over the whole line, with the arm's default reason. A
 * mapping carries `reason` (defaulting) and exactly one of: `commands`
 * as a list, a whole-line rule on each pattern; `commands` as a mapping,
 * each command pattern on its own paths (one command to many paths, one
 * rule per command); `paths` alone, a path rule on every command. A list
 * of commands beside a list of paths is refused, because it does not say
 * which command the paths belong to, and a rule naming neither is
 * refused rather than read as "every command".
 */
function parseRule(raw: unknown, where: string, defaultReason: string): CommandRule[] {
  if (typeof raw === 'string') {
    return [
      { reason: defaultReason, commands: stringList([raw], `${where}.commands`, 'a command') },
    ]
  }
  if (!isPlainObject(raw)) throw new Error(`${where} must be a command pattern or a mapping`)
  rejectUnknownKeys(raw, RULE_FIELDS, where)
  const reason = raw.reason ?? defaultReason
  if (typeof reason !== 'string') throw new Error(`${where}.reason must be a string`)
  const { commands, paths } = raw
  if (isPlainObject(commands)) {
    if (paths !== undefined && paths !== null) {
      throw new Error(`${where} maps each command to its paths, so it takes no paths of its own`)
    }
    return scopedRules(commands, reason, where)
  }
  const hasCommands = commands !== undefined && commands !== null
  const hasPaths = paths !== undefined && paths !== null
  if (hasCommands && hasPaths) {
    throw new Error(`${where} lists commands beside paths; map each command to its paths instead`)
  }
  if (!hasCommands && !hasPaths) throw new Error(`${where} names no command and no path`)
  return [
    {
      reason,
      commands: stringList(commands, `${where}.commands`, 'a command'),
      paths: stringList(paths, `${where}.paths`, 'a path'),
    },
  ]
}

function parseRules(raw: unknown, where: string, arm: 'ask' | 'deny'): readonly CommandRule[] {
  const fallback = arm === 'ask' ? DEFAULT_ASK_REASON : DEFAULT_DENY_REASON
  return asList(raw, `${where}.${arm}`, 'a list of rules').flatMap((entry, i) =>
    parseRule(entry, `${where}.${arm}[${String(i)}]`, fallback),
  )
}

function parseAllow(raw: unknown, where: string): readonly string[] | null {
  if (raw === undefined || raw === null) return null
  return stringList(raw, `${where}.allow`, 'a command')
}

/**
 * Validate a `paths:` block. `absolute` is the workspace and profile
 * tiers' rule (entries are virtual paths or name patterns); the mount
 * tier leaves it false because its entries are relative to the mount.
 */
export function parsePathsBlock(raw: unknown, where = 'paths', absolute = false): PathsBlock {
  const obj = asObject(raw, where)
  rejectUnknownKeys(obj, PATHS_FIELDS, where)
  const hide = stringList(obj.hide, `${where}.hide`, 'a path')
  if (absolute) requireAbsolute(hide, `${where}.hide`)
  return { hide }
}

export function parseVarsBlock(raw: unknown, where = 'vars'): VarsBlock {
  const obj = asObject(raw, where)
  rejectUnknownKeys(obj, VARS_FIELDS, where)
  return { hide: stringList(obj.hide, `${where}.hide`) }
}

export function parseCommandsBlock(raw: unknown, where = 'commands'): CommandsBlock {
  const obj = asObject(raw, where)
  rejectUnknownKeys(obj, COMMANDS_FIELDS, where)
  const ask = parseRules(obj.ask, where, 'ask')
  const deny = parseRules(obj.deny, where, 'deny')
  // This block is the workspace's or a profile's, never a mount's, so a
  // rule's paths are virtual paths: absolute, or name patterns.
  for (const rule of ask) requireAbsolute(rule.paths ?? [], `${where}.ask rule paths`)
  for (const rule of deny) requireAbsolute(rule.paths ?? [], `${where}.deny rule paths`)
  return { allow: parseAllow(obj.allow, where), ask, deny }
}

/** Validate a mount tier's `commands:` block (`ask` and `deny` only). */
export function parseMountCommandsBlock(raw: unknown, where = 'commands'): MountCommandsBlock {
  const obj = asObject(raw, where)
  rejectUnknownKeys(obj, MOUNT_COMMANDS_FIELDS, where)
  return { ask: parseRules(obj.ask, where, 'ask'), deny: parseRules(obj.deny, where, 'deny') }
}

/** Validate a `mounts.<prefix>.permissions` block. */
export function parseMountPermissions(raw: unknown, where = 'permissions'): MountPermissions {
  const obj = asObject(raw, where)
  rejectUnknownKeys(obj, MOUNT_PERMISSIONS_FIELDS, where)
  return {
    paths: obj.paths === undefined ? { hide: [] } : parsePathsBlock(obj.paths, `${where}.paths`),
    commands:
      obj.commands === undefined
        ? { ask: [], deny: [] }
        : parseMountCommandsBlock(obj.commands, `${where}.commands`),
  }
}

/** Validate the top-level `permissions:` block. */
export function parseWorkspacePermissions(
  raw: unknown,
  where = 'permissions',
): WorkspacePermissions {
  const obj = asObject(raw, where)
  rejectUnknownKeys(obj, WORKSPACE_PERMISSIONS_FIELDS, where)
  return {
    commands:
      obj.commands === undefined
        ? { allow: null, ask: [], deny: [] }
        : parseCommandsBlock(obj.commands, `${where}.commands`),
    paths:
      obj.paths === undefined ? { hide: [] } : parsePathsBlock(obj.paths, `${where}.paths`, true),
  }
}

/**
 * Normalize the `mounts` spellings a profile or `createSession` accepts:
 * a Map or Record of prefix to mode (names or `r`/`rw`/`rwx`), a string,
 * or an array of prefixes.
 */
export function parseProfileMounts(
  raw: unknown,
  where = 'mounts',
): ReadonlyMap<string, MountMode> | readonly string[] | null {
  if (raw === undefined || raw === null) return null
  if (typeof raw === 'string') return [normPrefix(raw)]
  if (Array.isArray(raw)) return stringList(raw, where).map(normPrefix)
  let entries: [unknown, unknown][]
  if (raw instanceof Map) entries = [...raw.entries()]
  else if (isPlainObject(raw)) entries = Object.entries(raw)
  else throw new Error(`${where} must be a mapping or a list of strings`)
  const modes = new Map<string, MountMode>()
  for (const [prefix, mode] of entries) {
    if (typeof prefix !== 'string') throw new Error(`${where} keys must be strings`)
    if (typeof mode !== 'string')
      throw new Error(`${where}[${prefix}] must be a mode name or alias`)
    modes.set(normPrefix(prefix), parseMountMode(mode))
  }
  return modes
}

/** Validate one profile (a `profiles.<name>` block, or an inline document). */
export function parseSessionProfile(raw: unknown, where = 'profile'): SessionProfile {
  const obj = asObject(raw, where)
  rejectUnknownKeys(obj, PROFILE_FIELDS, where)
  const out: {
    extends?: string | null
    cwd?: string | null
    env?: Readonly<Record<string, string>> | null
    mounts?: ReadonlyMap<string, string> | readonly string[] | null
    paths?: PathsBlock | null
    vars?: VarsBlock | null
    commands?: CommandsBlock | null
  } = {}
  if (obj.extends !== undefined && obj.extends !== null) {
    if (typeof obj.extends !== 'string') throw new Error(`${where}.extends must be a string`)
    out.extends = obj.extends
  }
  if (obj.cwd !== undefined && obj.cwd !== null) {
    if (typeof obj.cwd !== 'string') throw new Error(`${where}.cwd must be a string`)
    out.cwd = obj.cwd
  }
  if (obj.env !== undefined && obj.env !== null) {
    const env = asObject(obj.env, `${where}.env`)
    for (const [k, v] of Object.entries(env)) {
      if (typeof v !== 'string') throw new Error(`${where}.env.${k} must be a string`)
    }
    out.env = env as Record<string, string>
  }
  if (obj.mounts !== undefined && obj.mounts !== null) {
    out.mounts = parseProfileMounts(obj.mounts, `${where}.mounts`)
  }
  if (obj.paths !== undefined && obj.paths !== null)
    out.paths = parsePathsBlock(obj.paths, `${where}.paths`, true)
  if (obj.vars !== undefined && obj.vars !== null)
    out.vars = parseVarsBlock(obj.vars, `${where}.vars`)
  if (obj.commands !== undefined && obj.commands !== null)
    out.commands = parseCommandsBlock(obj.commands, `${where}.commands`)
  return out
}
